perf: identify identity columns once during init for postgres

kolaente opened the pull request:
This PR moves the identifying of identity columns from the insert query construction out and does it in the init function once for all tables. With this change, I observed init times for the testfixtures going down from 700ms to 30ms.

More details in the issue this is resolving.

Resolves https://github.com/go-testfixtures/testfixtures/issues/288

kolaente committed:
fix: only check for identity columns once, not every time when constructing an insert query

Resolves https://github.com/go-testfixtures/testfixtures/issues/288

slsyy commented on postgresql.go:
```diff
@@ -69,6 +69,10 @@ func (h *postgreSQL) init(db *sql.DB) error {
 	}
 
 	h.tablesHasIdentityColumn = make(map[string]bool)
+	err := h.buildTableHasIdentityColumn(db)
```
please make it simliar to functions above, so:
* under grp.Go to get parallel execution
* function return a value so `h.tablesHasIdentityColumn = h.buildTableHasIdentityColumn(db)` for simpler data flow

kolaente replied:
Done

slsyy commented on postgresql.go:
```diff
@@ -442,34 +442,42 @@ func (h *postgreSQL) buildInsertSQL(q shared.Queryable, tableName string, column
+	const query = `SELECT table_name, COUNT(*) AS identity_column_count
+    FROM information_schema.columns
+    WHERE
+      table_schema NOT IN ('pg_catalog', 'information_schema', 'crdb_internal') AND
```
is those new checks necessary?

kolaente replied:
Not necessarily, but we don't need to store that information, don't we?

slsyy replied:
Yes, true

slsyy commented on postgresql.go:
```diff
@@ -442,34 +442,42 @@ func (h *postgreSQL) buildInsertSQL(q shared.Queryable, tableName string, column
+}
+
+func (h *postgreSQL) buildTableHasIdentityColumn(q shared.Queryable) error {
 	defer h.tablesHasIdentityColumnMutex.Unlock()
```
i think the mutex is not needed anymore

kolaente replied:
Right, when returning the data. I've removed it.

slsyy commented:
Please change the title to `perf: identify identity columns once during init for postgres`. It is not `fix`, because it works (albeit slowly)

kolaente commented on postgresql.go:
```diff
@@ -442,34 +439,41 @@ func (h *postgreSQL) buildInsertSQL(q shared.Queryable, tableName string, column
-	defer h.tablesHasIdentityColumnMutex.Unlock()
-	h.tablesHasIdentityColumnMutex.Lock()
+func (h *postgreSQL) tableHasIdentityColumn(tableName string) bool {
+	tableName = strings.Trim(tableName, `"`)
```
I found this to be interesting, and it seems like a quirk. Is it intended the table name has quotes around it?

slsyy replied:
I am not sure (i will check it later). As I remember we add quotes in a code somewhere at the begining of the process, so hand-written queries do not struggle with some crazy table names like names with spaces and so on

kolaente replied:
okay, I guess then it's safe to remove them here?

slsyy replied:
I am worried about situation, when you have the same table name but in a different `table_schema`

Anyway your change does not change anything in that regard, so let's keep it

kolaente commented:
When I run only the failing test locally, it passes, when I run all tests locally, the same test fails. Any ideas?

slsyy commented:
@kolaente nope. I will check and test it manually later, if you don't find a solution

slsyy commented:
@kolaente I got the same results as CI. I run it with `go-task docker:build &&  go-task docker:test`

kolaente commented:
It seems like this is caused by the table name being `"public"."posts"` and not `posts` (which is stored in the map). Will check how to fix this.

kolaente committed:
fix: handle database.table table names correctly

kolaente commented:
Fixed in c9de039f867e36e77923f9e6d0191488e461ba78 - there was similar code in the version I'm replacing with this PR, so I just adjusted it like it was before.

kolaente commented:
Thank you very much!

Can we get a release for this?

slsyy commented:
@kolaente can you use the hash version for now? I won't be available for the next week, so i am a little bit scared to release it now without any support

kolaente commented:
Sure, I can do that. Many thanks!

slsyy commented:
@kolaente released in https://github.com/go-testfixtures/testfixtures/releases/tag/v3.16.0

kolaente commented:
Thanks!
